### ISSUE TITLE
Oracle Queries fail - Unable to display Raw Data

### ISSUE THREAD

**Issue by 6sicuro** [metabase/metabase#3568]
- Browser: Chrome 53.0.2785.143 m
- OS: Windows 7
- Oracle 10.2 with Ojdbc7
- Metabase version 0.20.0
- Metabase hosting environment: Linux Debian 6
- Metabase internal database: H2

As a issue subject I have a problem during "new question" when my request is for raw data . Indipendently from table’s rows/columns count I get always the error attached.

I've not tested the issue over others DB.

![image](https://cloud.githubusercontent.com/assets/1751529/19384625/74eeb42c-920a-11e6-9814-1aea52b07e93.png)

**Comment by hugocar** [comment]
(I m not from metabase staff)

Have you tried running "SELECT * FROM your_table LIMIT 2000" in an other software ? like directly in the console terminal or Workbench ?

**Comment by camsaul** [comment]
@6sicuro, @hugocar's suggestion sounds like a good thing to try. See if that's working (I would assume it is, because Metabase runs a test query that has to complete successfully before it lets you save a new Database).

If that works can you post the SQL that gets generated? You can find that by clicking the "View the SQL" button in the top-right near the "SAVE" button. That will help debug your issue.

**Comment by camsaul** [comment]
@salsakran don't think this is an issue with the Query Builder because the query processor never actually returns results in this case and it's just a plain Raw data query.

**Comment by DiegoUsai** [comment]
@camsaul @hugocar "select *"query works using other software and works on metabase too when I wrote it manually in sql mode..
The error looks like is a sql construction error (ORA-00933: SQL command not properly ended).
In log file I found this query 

> SELECT ... FROM ... **OFFSET 0 ROWS FETCH NEXT 2000 ROWS ONLY**

The bold part looks like a non valid oracle commad (I've tried in other software), the correct way to limit oracle query is **WHERE ROWNUM<=2000**

**Comment by camsaul** [comment]
@DiegoUsai after looking into this it looks like the `FETCH NEXT n ROWS ONLY` syntax was about 3½ years ago with Oracle 12.

Does Oracle 10.x support any other equivalent syntax? I'm happy to tweak the SQL generating code if there's something that will work for older versions of Oracle as well.

**Comment by DiegoUsai** [comment]
@camsaul Oracle 10.x does not support `FETCH NEXT n ROWS ONLY`but only `WHERE ROWNUM<=n , I test it.
I hope this help

**Comment by Charles-Go** [comment]
I'm having the same issue. Wrapping the query with `SELECT * FROM (original_query) WHERE ROWNUM<=2000` should work as @6sicuro said

**Comment by camsaul** [comment]
@DiegoUsai @Charles-Go 

Ok to make sure I understand this correctly, something like this won't work:

```sql
SELECT *
FROM employees
WHERE RONUM <= 10
ORDER BY employee_id;
```

instead, it has to be:

```sql
SELECT * 
FROM (
    SELECT * 
    FROM employees 
    ORDER BY employee_id
)
WHERE ROWNUM < 10;
```

is that correct?

**Comment by camsaul** [comment]
@DiegoUsai @Charles-Go one more question. If I understand this correctly, there's no way to do the equivalent of `OFFSET` in Oracle before version 12, correct?

**Comment by Charles-Go** [comment]
@camsaul Yes that is correct. "A query with WHERE ROWNUM = 5 or WHERE ROWNUM > 5 doesn't make sense. This is because a ROWNUM value is assigned to a row during the predicate evaluation and gets incremented only after a row passes the WHERE clause." http://www.oracle.com/technetwork/issue-archive/2006/06-sep/o56asktom-086197.html

**Comment by Charles-Go** [comment]
Do you want to be doing pagination with the OFFSET? If so, the link I shared explains a way for it. (sadly from way back)

**Comment by camsaul** [comment]
So to do the equivalent of `OFFSET` I would have to do

```sql
SELECT *
FROM (
    SELECT *, ROWNUM rnum
    FROM (
        SELECT *
        FROM employees
        ORDER BY employee_id
    )
    WHERE ROWNUM <= 150
)
WHERE rnum >= 100;
```

😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂

**Comment by camsaul** [comment]
Ok I figured this all out. I have the Oracle driver set up to generate SQL like this:

Postgres:
```sql
SELECT "x", "y", "z"
FROM checkins
OFFSET 100
LIMIT 10
```

Oracle:
```sql
SELECT *
FROM (
    SELECT "__table__".*, ROWNUM AS "__rownum__"
    FROM (
        SELECT "x", "y", "z"
        FROM "checkins"
    ) "__table__"
    WHERE ROWNUM <= 100
)
WHERE "__rownum__" > 50
```

Once again, 😂  😹

**Comment by tlrobinson** [comment]
@camsaul Maybe you're already do this, but can we make sure to only use the ugly SQL if an offset is actually needed?

**Comment by camsaul** [comment]
@tlrobinson Yup we're doing that. If an offset isn't needed it only needs to do one nested query instead of two

**Comment by camsaul** [comment]
Fixed by #3662
